packages/react-ui-core/src/AutoSuggestField/AutoSuggestField.js

```diff
@@ -0,0 +1,205 @@
+import { Dropdown } from '../Dropdown'
+import { Menu } from '../Menu'
+import InputAnchor from './InputAnchor'
+import { Button } from '../Button'
+
+const ENTER = 13
+const ESCAPE = 27
+
+@themed(/^AutoSuggestField/, {
+  pure: true,
+})
+
+export default class AutoSuggestField extends Component {
+  static propTypes = {
+    className: PropTypes.string,
+    theme: PropTypes.object,
+    onInput: PropTypes.func,
+    suggestions: PropTypes.node,
+    resetButton: PropTypes.oneOfType([
+      PropTypes.node,
+      PropTypes.func,
+      PropTypes.object,
+    ]),
+    onAfterReset: PropTypes.func,
+    submitButton: PropTypes.oneOfType([
+      PropTypes.node,
+      PropTypes.func,
+      PropTypes.object,
+    ]),
+    onSubmit: PropTypes.func,
+    onSelection: PropTypes.func,
+    value: PropTypes.string,
+    placeholder: PropTypes.string,
+    suggestionValueSelector: PropTypes.func,
+    onSelectionHover: PropTypes.func,
+    submitOnSelection: PropTypes.bool,
+  }
+
+  static defaultProps = {
+    theme: {},
+    value: '',
+    onAfterReset: () => {},
+    onSubmit: () => {},
+    onInput: () => {},
+    onSelection: () => {},
+    submitOnSelection: true,
+  }
+
+  constructor(props) {
+    super(props)
+    const { value } = this.props
+    this.onInput = debounce(this.props.onInput, 300, { leading: false, trailing: true })
+    this.state = {
+      value,
+      visible: true,
+    }
+    this.onInput(value)
+  }
+
+  componentDidMount() {
+    document.addEventListener('keydown', this.onKeyDown)
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('keydown', this.onKeyDown)
+  }
+
+  @autobind
+  onKeyDown(event) {
+    const { suggestions } = this.props
+    const { visible } = this.state
+
+    switch (event.keyCode) {
+      case ESCAPE:
+        event.preventDefault()
+        this.reset()
+        break
+      case ENTER:
+        if (!visible || !suggestions || suggestions.length < 1) {
+          event.preventDefault()
+          this.handleOnSubmit(this.state.value)
+        }
+        break
+      default:
+    }
+  }
+
+  @autobind
+  handleInput(event) {
+    const value = event.target.value
+    this.setState({ value })
+    this.onInput(value)
+  }
+
+  @autobind
+  handleOnSubmit() {
+    this.props.onSubmit(this.state.value)
+    this.setState({ visible: false })
+  }
+
+  @autobind
+  handleSuggestionSelection(value) {
+    const {
+      onSubmit,
+      onSelection,
+      submitOnSelection,
+      onInput,
+    } = this.props
+
+    this.setState({
+      value,
+      visible: false,
+    })
+
+    onInput(value)
+    onSelection(value)
+    if (submitOnSelection) onSubmit(value)
+  }
+
+  @autobind
+  handleVisibilityChange(visible) {
+    this.setState({
+      visible,
+    })
+  }
+
+  @autobind
+  reset() {
+    this.setState({ value: '', visible: false })
+    this.props.onAfterReset()
+  }
+
+  renderButton(button, handleClick, dataTid, props) {
+    if (!button) return undefined
```

Comment:
`null` is the appropriate return value here.  React needs either a node or null to render with.

Reply:
Hmm.. I guess they've since changed this: https://reactjs.org/docs/jsx-in-depth.html#booleans-null-and-undefined-are-ignored

`undefined` is fine here.

Reply:
👍